### PR TITLE
Alerting: Use URLs instead of tokens for images in notifications

### PR DESCRIPTION
**Pull request by santihernandezc**
### Description

This PR replaces tokens for URLs in `ImageTokenAnnotation`, which will allow us to add support for images in the Grafana Cloud Alertmanager.

There are two types of URLs:
- Public ones, pointing to an external image store.
- Local ones, prefixed with `file://`

### Changes

- The `ImageTokenAnnotation` is now the image's URL, public or local.
- URLs replace tokens as unique identifiers for images, so integrations can still query the DB for images using the value extracted from `ImageTokenAnnotation`.
- In cases of no public URL, a URL is generated using the path to the image on disk and prefixing it with `file://`.
- Instead of adding a `*models.Image` in state changes, the state manager just adds a string, which is everything we were using out of that struct anyway.

### CODE REVIEW

**Comment by santihernandezc** [comment]
@grobinson-grafana Integrations would still query the database, where they would potentially get both the `URL` and the `Path` fields, so we can still have both.
This change is for the Cloud AM to get a link to the image without access to the database.

File: pkg/services/ngalert/notifier/images.go [hunk 1]
```diff
@@ -20,21 +19,21 @@ func newImageStore(store store.ImageStore) images.ImageStore {
...
-	var result *images.Image
-	if image != nil {
-		result = &images.Image{
-			Token:     image.Token,
-			Path:      image.Path,
-			URL:       image.URL,
-			CreatedAt: image.CreatedAt,
-		}
+
+	if strings.HasPrefix(image.URL, "file://") {
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/images.go:30]: we need to explain why we do it here. Please leave a comment that clarifies this.

File: pkg/services/sqlstore/migrations/ualert/tables.go [hunk 2]
```diff
@@ -488,6 +488,9 @@ func addAlertImageMigrations(mg *migrator.Migrator) {
 	mg.AddMigration("support longer URLs in alert_image table", migrator.NewRawSQLMigration("").
 		Postgres("ALTER TABLE alert_image ALTER COLUMN url TYPE VARCHAR(2048);").
 		Mysql("ALTER TABLE alert_image MODIFY url VARCHAR(2048) NOT NULL;"))
+
+	mg.AddMigration("drop unique `token` index on alert_image", migrator.NewDropIndexMigration(imageTable, imageTable.Indices[0]))
+	mg.AddMigration("add unique index on `url` to alert_image", migrator.NewAddIndexMigration(imageTable, &migrator.Index{Type: migrator.UniqueIndex, Cols: []string{"url"}}))
```
Comment (alexweav) [inline pkg/services/sqlstore/migrations/ualert/tables.go:493]: was URL already likely to be unique?

adding a unique index will fail if there are existing records that already break the uniqueness constraint. in what situations can existing images have the same URL? is this possible today? if so, do we need to clean out conflicting records before adding this?

if URL was guaranteed based on token before this change, like `.../path/to/some/<token>` it's safe. What do the URLs look like in practice?

Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/tables.go:493]: + that will be broken when Grafana is configured to not upload anywhere and keep it only on disc (i.e. URL is empty)

Comment (santihernandezc) [reply pkg/services/sqlstore/migrations/ualert/tables.go:493]: @alexweav thanks for pointing this out! The index indeed causes an error in the migrations.

**Comment by santihernandezc** [comment]
Closing in favor of https://github.com/grafana/grafana/pull/66804

**Pull request closed without merging**